aggsender/aggsender.go

```diff
@@ -416,24 +428,120 @@ func (a *AggSender) sendCertificate(ctx context.Context) (*agglayertypes.Certifi
 	return certificate, nil
 }
 
-// callValidator calls the validator to validate the certificate
-func (a *AggSender) callValidator(
+// pollValidators committee signers to validate the certificate and gather signatures
+func (a *AggSender) pollValidators(
 	ctx context.Context,
+	validators []types.CertificateValidateAndSigner,
+	signaturesThreshold uint32,
 	certificate *agglayertypes.Certificate,
-	lastL2BlockInCert uint64) ([]byte, error) {
-	if a.validator == nil {
+	lastL2BlockInCert uint64,
+) ([][]byte, error) {
+	if len(validators) == 0 {
+		a.log.Warnf("skipping certificate validation, because there are no validators configured")
+
 		return nil, nil
 	}
+
 	a.log.Infof("delegating certificate validation: %s", certificate.Brief())
-	validatorSignature, err := a.validator.ValidateAndSignCertificate(ctx, certificate, lastL2BlockInCert)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	sigCh := make(chan []byte, len(validators))
+	errCh := make(chan error, len(validators))
+
+	g, gctx := errgroup.WithContext(ctx)
+
+	for _, v := range validators {
+		g.Go(func() error {
+			status, err := v.HealthCheck(ctx)
+			if err != nil {
+				a.log.Warnf("error checking validator health (URL=%s): %v", v.URL(), err)
+			}
+
+			if !status.IsHealthy() {
+				a.log.Warnf("validator status (URL=%s) is not healthy: %s", v.URL(), status.String())
+				return nil // skip the unhealthy validator
+			}
+
+			a.log.Infof("validator health check (URL=%s): %s", v.URL(), status.String())
+
+			sig, err := v.ValidateAndSignCertificate(gctx, certificate, lastL2BlockInCert)
+			if err != nil {
+				a.log.Errorf("validator %v failed to validate the certificate: %v", v, err)
+				errCh <- err
+				return nil // don't fail the group
+			}
+
+			select {
+			case sigCh <- sig:
+				if uint32(len(sigCh)) >= signaturesThreshold {
+					cancel() // early stop, because threshold is reached
+				}
+			case <-gctx.Done():
+				return nil
+			}
+			return nil
+		})
+	}
+
+	// Wait for all goroutines to finish in background
+	go func() {
+		_ = g.Wait()
+		close(sigCh)
+		close(errCh)
+	}()
+
+	signatures := make([][]byte, 0, len(validators))
+	for sig := range sigCh {
```

Comment:
Hmm, the logic with multiple channels is really complex to me. Maybe we can just add a `mutex` and use it to write to signature and errors arrays. It will also decrease the lines of code in the function. WDYT?

Reply:
Kept the channel, but simplified a bit, by introducing the struct that holds both signature and error. Let me know if you still think we should simplify it https://github.com/agglayer/aggkit/pull/877/commits/80268b2d8956aa6b376701c8b795ebd9964da0d7